tasks/static_quality_gates/lib/gates_lib.py

```diff
@@ -262,3 +273,350 @@ def generate_metric_reports(self, ctx, filename="static_gate_report.json", branc
                 f"aws s3 cp --only-show-errors --region us-east-1 --sse AES256 {filename} {self.S3_REPORT_PATH}/{CI_COMMIT_SHA}/{filename}",
                 hide="stdout",
             )
+
+
+class StaticQualityGateFailed(Exception):
+    """
+    Exception raised when a static quality gate fails
+    """
+
+    def __init__(self, message: str):
+        self.message = color_message(message, "red")
+        super().__init__(self.message)
+
+
+class StaticQualityGate:
+    """
+    Base class for all static quality gates
+    that contains the common attributes for all static quality gates
+    """
+
+    gate_name: str
+    arch: str
+    os: str
+    max_on_wire_size: int
+    max_on_disk_size: int
+    artifact_on_disk_size: int
+    artifact_on_wire_size: int
+    artifact_path: str  # Path to the artifact to be used for the gate
+
+    def __init__(self, gate_name: str, gate_max_size_values: dict):
+        self.gate_name = gate_name
+        self.max_on_wire_size = read_byte_input(gate_max_size_values["max_on_wire_size"])
+        self.max_on_disk_size = read_byte_input(gate_max_size_values["max_on_disk_size"])
+        self._set_arch()
+
+    def _set_arch(self):
+        if "amd64" in self.gate_name:
+            self.arch = "amd64"
+        elif "arm64" in self.gate_name:
+            self.arch = "arm64"
+        elif "armhf" in self.gate_name:
+            self.arch = "armhf"
+        elif "msi" in self.gate_name:
+            # MSI packages are always amd64 (x86_64) on Windows
+            self.arch = "amd64"
+        else:
+            raise ValueError(f"Unknown arch for gate: {self.gate_name}")
+
+    def print_results(self) -> None:
+        """
+        Print the results of the gate
+        in case of success
+        """
+        from .static_quality_gates_reporter import QualityGateOutputFormatter
+
+        QualityGateOutputFormatter.print_enhanced_gate_result(
+            self.gate_name,
+            self.artifact_path,
+            self.artifact_on_wire_size,
+            self.max_on_wire_size,
+            self.artifact_on_disk_size,
+            self.max_on_disk_size,
+        )
+
+    @abstractmethod
+    def _measure_on_disk_and_on_wire_size(self, ctx: Context) -> None:
+        """
+        Measure the size of the artifact on disk and on wire
+        """
+        raise NotImplementedError("This method should be implemented by the subclass")
```

Comment:
There's a problem with this in that it's not clear what this method is supposed to do. We can infer from the the code that it's expected to populate some internal state, but that's not documented here and feels brittle regardless.

I think this should return the values rather than modify state. That's also somewhat related to why I preferred composition over inheritance here.